godot-macros/src/util/mod.rs

```diff
@@ -303,3 +315,92 @@ pub fn venial_parse_meta(
+
+pub fn make_func_name_constants(funcs: &[FuncDefinition], class_name: &Ident) -> Vec<TokenStream> {
+    funcs
+        .iter()
+        .map(|func| {
+            // The constant needs the same #[cfg] attribute(s) as the function, so that it is only active if the function is also active.
+            let cfg_attributes = extract_cfg_attrs(&func.external_attributes)
+                .into_iter()
+                .collect::<Vec<_>>();
+
+            make_func_name_constant(
+                class_name,
+                &func.signature_info.method_name,
+                func.registered_name.as_ref(),
+                &cfg_attributes,
+            )
+        })
+        .collect()
+}
+
+/// Funcs can be renamed with `#[func(rename=new_name) fn f();`.
+/// To be able to access the renamed function name at a later point, it is saved in a string constant.
+pub fn make_func_name_constant(
+    class_name: &Ident,
+    func_name: &Ident,
+    registered_name: Option<&String>,
+    attributes: &[&Attribute],
+) -> TokenStream {
+    let const_name = format_func_name_constant_name(class_name, func_name);
+    let const_value = match &registered_name {
+        Some(renamed) => renamed.to_string(),
+        None => func_name.to_string(),
+    };
+
+    let doc_comment =
+        format!("The Rust function `{func_name}` is registered with Godot as `{const_value}`.");
+
+    quote! {
+        #(#attributes)*
+        #[doc = #doc_comment]
+        #[doc(hidden)]
+        #[allow(non_upper_case_globals)]
+        pub const #const_name: &str  = #const_value;
+    }
+}
+
+/// Converts "path::class" to "path::new_class".
+pub fn replace_class_in_path(path: Path, new_class: Ident) -> Path {
+    match path.segments.as_slice() {
+        // Can't happen, you have at least one segment (the class name).
+        [] => panic!("unexpected: empty path"),
+
+        [_single] => Path {
+            segments: vec![PathSegment {
+                ident: new_class,
+                generic_args: None,
+                tk_separator_colons: None,
+            }],
+        },
+
+        [path @ .., _last] => {
+            let mut segments = vec![];
+            segments.extend(path.iter().cloned());
+            segments.push(PathSegment {
+                ident: new_class,
+                generic_args: None,
+                tk_separator_colons: Some([
+                    Punct::new(':', Spacing::Joint),
+                    Punct::new(':', Spacing::Alone),
+                ]),
+            });
+            Path { segments }
+        }
+    }
+}
+
+/// Returns the name of the constant that will be autogenerated.
+pub fn format_func_name_constant_name(_class_name: &Ident, func_name: &Ident) -> Ident {
+    format_ident!("{func_name}")
+}
+
+/// Returns the name of the dummy struct that's used as container for all function name constants.
+pub fn format_func_name_struct_name(class_name: &Ident) -> Ident {
+    format_ident!("__gdext_{class_name}_Functions")
```

Comment:
why is `Functions` capitalized :thinking:?

Reply:
I think that's OK, it's a struct name after all. The prefix `__gdext_` doesn't follow any conventions, and the `_` before `Functions` makes it easier to tell apart from the class name.

Reply:
makes sense